File: src/org/joml/Matrix2f.java
package org.joml;

import java.text.NumberFormat;

@SuppressWarnings("unused")
public class Matrix2f {

    public float m00;
    public float m01;
    public float m10;
    public float m11;

    public Matrix2f() {
        this.m00 = 1.0F;
        this.m11 = 1.0F;
    }

    public Matrix2f(Matrix2f mat) {
        if (mat instanceof Matrix2f) {
            MemUtil.INSTANCE.copy((Matrix2f) mat, this);
        } else {
            this.setMatrix2f(mat);
        }

    }

    public Matrix2f(Matrix3f mat) {
        this.m00 = mat.m00;
        this.m01 = mat.m01;
        this.m10 = mat.m10;
        this.m11 = mat.m11;
    }

    public Matrix2f(float m00, float m01, float m10, float m11) {
        this.m00 = m00;
        this.m01 = m01;
        this.m10 = m10;
        this.m11 = m11;
    }

    public Matrix2f(Vector2f col0, Vector2f col1) {
        this.m00 = col0.x;
        this.m01 = col0.y;
        this.m10 = col1.x;
        this.m11 = col1.y;
    }

    public float m00() {
        return this.m00;
    }

    public float m01() {
        return this.m01;
    }

    public float m10() {
        return this.m10;
    }

    public float m11() {
        return this.m11;
    }

    public Matrix2f m00(float m00) {
        this.m00 = m00;
        return this;
    }

    public Matrix2f m01(float m01) {
        this.m01 = m01;
        return this;
    }

    public Matrix2f m10(float m10) {
        this.m10 = m10;
        return this;
    }

    public Matrix2f m11(float m11) {
        this.m11 = m11;
        return this;
    }

    Matrix2f _m00(float m00) {
        this.m00 = m00;
        return this;
    }

    Matrix2f _m01(float m01) {
        this.m01 = m01;
        return this;
    }

    Matrix2f _m10(float m10) {
        this.m10 = m10;
        return this;
    }

    Matrix2f _m11(float m11) {
        this.m11 = m11;
        return this;
    }

    public Matrix2f set(Matrix2f m) {
        if (m instanceof Matrix2f) {
            MemUtil.INSTANCE.copy((Matrix2f) m, this);
        } else {
            this.setMatrix2f(m);
        }

        return this;
    }

    private void setMatrix2f(Matrix2f mat) {
        this.m00 = mat.m00;
        this.m01 = mat.m01;
        this.m10 = mat.m10;
        this.m11 = mat.m11;
    }

    public Matrix2f set(Matrix3x2f m) {
        this.m00 = m.m00;
        this.m01 = m.m01;
        this.m10 = m.m10;
        this.m11 = m.m11;
        return this;
    }

    public Matrix2f set(Matrix3f m) {
        this.m00 = m.m00;
        this.m01 = m.m01;
        this.m10 = m.m10;
        this.m11 = m.m11;
        return this;
    }

    private void setMatrix3f(Matrix3f mat) {
        this.m00 = mat.m00;
        this.m01 = mat.m01;
        this.m10 = mat.m10;
        this.m11 = mat.m11;
    }

    public Matrix2f mul(Matrix2f right) {
        return this.mul(right, this);
    }

    public Matrix2f mul(Matrix2f right, Matrix2f dest) {
        float nm00 = this.m00 * right.m00 + this.m10 * right.m01;
        float nm01 = this.m01 * right.m00 + this.m11 * right.m01;
        float nm10 = this.m00 * right.m10 + this.m10 * right.m11;
        float nm11 = this.m01 * right.m10 + this.m11 * right.m11;
        dest.m00 = nm00;
        dest.m01 = nm01;
        dest.m10 = nm10;
        dest.m11 = nm11;
        return dest;
    }

    public Matrix2f mulLocal(Matrix2f left) {
        return this.mulLocal(left, this);
    }

    public Matrix2f mulLocal(Matrix2f left, Matrix2f dest) {
        float nm00 = left.m00 * this.m00 + left.m10 * this.m01;
        float nm01 = left.m01 * this.m00 + left.m11 * this.m01;
        float nm10 = left.m00 * this.m10 + left.m10 * this.m11;
        float nm11 = left.m01 * this.m10 + left.m11 * this.m11;
        dest.m00 = nm00;
        dest.m01 = nm01;
        dest.m10 = nm10;
        dest.m11 = nm11;
        return dest;
    }

    public Matrix2f set(float m00, float m01, float m10, float m11) {
        this.m00 = m00;
        this.m01 = m01;
        this.m10 = m10;
        this.m11 = m11;
        return this;
    }

    public Matrix2f set(float[] m) {
        MemUtil.INSTANCE.copy(m, 0, this);
        return this;
    }

    public Matrix2f set(Vector2f col0, Vector2f col1) {
        this.m00 = col0.x;
        this.m01 = col0.y;
        this.m10 = col1.x;
        this.m11 = col1.y;
        return this;
    }

    public float determinant() {
        return this.m00 * this.m11 - this.m10 * this.m01;
    }

    public Matrix2f invert() {
        return this.invert(this);
    }

    public Matrix2f invert(Matrix2f dest) {
        float s = 1.0F / this.determinant();
        float nm00 = this.m11 * s;
        float nm01 = -this.m01 * s;
        float nm10 = -this.m10 * s;
        float nm11 = this.m00 * s;
        dest.m00 = nm00;
        dest.m01 = nm01;
        dest.m10 = nm10;
        dest.m11 = nm11;
        return dest;
    }

    public Matrix2f transpose() {
        return this.transpose(this);
    }

    public Matrix2f transpose(Matrix2f dest) {
        dest.set(this.m00, this.m10, this.m01, this.m11);
        return dest;
    }

    public String toString() {
        String str = this.toString(Options.NUMBER_FORMAT);
        StringBuilder res = new StringBuilder();
        int eIndex = Integer.MIN_VALUE;

        for (int i = 0; i < str.length(); ++i) {
            char c = str.charAt(i);
            if (c == 'E') {
                eIndex = i;
            } else {
                if (c == ' ' && eIndex == i - 1) {
                    res.append('+');
                    continue;
                }

                if (Character.isDigit(c) && eIndex == i - 1) {
                    res.append('+');
                }
            }

            res.append(c);
        }

        return res.toString();
    }

    public String toString(NumberFormat formatter) {
        return Runtime.format(this.m00, formatter) + " " + Runtime.format(this.m10, formatter) + "\n" + Runtime.format(this.m01, formatter) + " " + Runtime.format(this.m11, formatter) + "\n";
    }

    public Matrix2f get(Matrix2f dest) {
        return dest.set((Matrix2f) this);
    }

    public Matrix3x2f get(Matrix3x2f dest) {
        return dest.set(this);
    }

    public Matrix3f get(Matrix3f dest) {
        return dest.set(this);
    }

    public float getRotation() {
        return Math.atan2(this.m01, this.m11);
    }

    public float[] get(float[] arr, int offset) {
        MemUtil.INSTANCE.copy(this, arr, offset);
        return arr;
    }

    public float[] get(float[] arr) {
        return this.get(arr, 0);
    }

    public Matrix2f zero() {
        MemUtil.INSTANCE.zero(this);
        return this;
    }

    public Matrix2f identity() {
        MemUtil.INSTANCE.identity(this);
        return this;
    }

    public Matrix2f scale(Vector2f xy, Matrix2f dest) {
        return this.scale(xy.x, xy.y, dest);
    }

    public Matrix2f scale(Vector2f xy) {
        return this.scale(xy.x, xy.y, this);
    }

    public Matrix2f scale(float x, float y, Matrix2f dest) {
        dest.m00 = this.m00 * x;
        dest.m01 = this.m01 * x;
        dest.m10 = this.m10 * y;
        dest.m11 = this.m11 * y;
        return dest;
    }

    public Matrix2f scale(float x, float y) {
        return this.scale(x, y, this);
    }

    public Matrix2f scale(float xy, Matrix2f dest) {
        return this.scale(xy, xy, dest);
    }

    public Matrix2f scale(float xy) {
        return this.scale(xy, xy);
    }

    public Matrix2f scaleLocal(float x, float y, Matrix2f dest) {
        dest.m00 = x * this.m00;
        dest.m01 = y * this.m01;
        dest.m10 = x * this.m10;
        dest.m11 = y * this.m11;
        return dest;
    }

    public Matrix2f scaleLocal(float x, float y) {
        return this.scaleLocal(x, y, this);
    }

    public Matrix2f scaling(float factor) {
        MemUtil.INSTANCE.zero(this);
        this.m00 = factor;
        this.m11 = factor;
        return this;
    }

    public Matrix2f scaling(float x, float y) {
        MemUtil.INSTANCE.zero(this);
        this.m00 = x;
        this.m11 = y;
        return this;
    }

    public Matrix2f scaling(Vector2f xy) {
        return this.scaling(xy.x, xy.y);
    }

    public Matrix2f rotation(float angle) {
        float sin = Math.sin(angle);
        float cos = Math.cosFromSin(sin, angle);
        this.m00 = cos;
        this.m01 = sin;
        this.m10 = -sin;
        this.m11 = cos;
        return this;
    }

    public Vector2f transform(Vector2f v) {
        return v.mul(this);
    }

    public Vector2f transform(Vector2f v, Vector2f dest) {
        v.mul(this, dest);
        return dest;
    }

    public Vector2f transform(float x, float y, Vector2f dest) {
        dest.set(this.m00 * x + this.m10 * y, this.m01 * x + this.m11 * y);
        return dest;
    }

    public Vector2f transformTranspose(Vector2f v) {
        return v.mulTranspose(this);
    }

    public Vector2f transformTranspose(Vector2f v, Vector2f dest) {
        v.mulTranspose(this, dest);
        return dest;
    }

    public Vector2f transformTranspose(float x, float y, Vector2f dest) {
        dest.set(this.m00 * x + this.m01 * y, this.m10 * x + this.m11 * y);
        return dest;
    }

    public Matrix2f rotate(float angle) {
        return this.rotate(angle, this);
    }

    public Matrix2f rotate(float angle, Matrix2f dest) {
        float s = Math.sin(angle);
        float c = Math.cosFromSin(s, angle);
        float nm00 = this.m00 * c + this.m10 * s;
        float nm01 = this.m01 * c + this.m11 * s;
        float nm10 = this.m10 * c - this.m00 * s;
        float nm11 = this.m11 * c - this.m01 * s;
        dest.m00 = nm00;
        dest.m01 = nm01;
        dest.m10 = nm10;
        dest.m11 = nm11;
        return dest;
    }

    public Matrix2f rotateLocal(float angle) {
        return this.rotateLocal(angle, this);
    }

    public Matrix2f rotateLocal(float angle, Matrix2f dest) {
        float s = Math.sin(angle);
        float c = Math.cosFromSin(s, angle);
        float nm00 = c * this.m00 - s * this.m01;
        float nm01 = s * this.m00 + c * this.m01;
        float nm10 = c * this.m10 - s * this.m11;
        float nm11 = s * this.m10 + c * this.m11;
        dest.m00 = nm00;
        dest.m01 = nm01;
        dest.m10 = nm10;
        dest.m11 = nm11;
        return dest;
    }

    public Vector2f getRow(int row, Vector2f dest) throws IndexOutOfBoundsException {
        switch (row) {
            case 0:
                dest.x = this.m00;
                dest.y = this.m10;
                break;
            case 1:
                dest.x = this.m01;
                dest.y = this.m11;
                break;
            default:
                throw new IndexOutOfBoundsException();
        }

        return dest;
    }

    public Matrix2f setRow(int row, Vector2f src) throws IndexOutOfBoundsException {
        return this.setRow(row, src.x, src.y);
    }

    public Matrix2f setRow(int row, float x, float y) throws IndexOutOfBoundsException {
        switch (row) {
            case 0:
                this.m00 = x;
                this.m10 = y;
                break;
            case 1:
                this.m01 = x;
                this.m11 = y;
                break;
            default:
                throw new IndexOutOfBoundsException();
        }

        return this;
    }

    public Vector2f getColumn(int column, Vector2f dest) throws IndexOutOfBoundsException {
        switch (column) {
            case 0:
                dest.x = this.m00;
                dest.y = this.m01;
                break;
            case 1:
                dest.x = this.m10;
                dest.y = this.m11;
                break;
            default:
                throw new IndexOutOfBoundsException();
        }

        return dest;
    }

    public Matrix2f setColumn(int column, Vector2f src) throws IndexOutOfBoundsException {
        return this.setColumn(column, src.x, src.y);
    }

    public Matrix2f setColumn(int column, float x, float y) throws IndexOutOfBoundsException {
        switch (column) {
            case 0:
                this.m00 = x;
                this.m01 = y;
                break;
            case 1:
                this.m10 = x;
                this.m11 = y;
                break;
            default:
                throw new IndexOutOfBoundsException();
        }

        return this;
    }

    public float get(int column, int row) {
        switch (column) {
            case 0:
                switch (row) {
                    case 0:
                        return this.m00;
                    case 1:
                        return this.m01;
                    default:
                        throw new IndexOutOfBoundsException();
                }
            case 1:
                switch (row) {
                    case 0:
                        return this.m10;
                    case 1:
                        return this.m11;
                }
        }

        throw new IndexOutOfBoundsException();
    }

    public Matrix2f set(int column, int row, float value) {
        switch (column) {
            case 0:
                switch (row) {
                    case 0:
                        this.m00 = value;
                        return this;
                    case 1:
                        this.m01 = value;
                        return this;
                    default:
                        throw new IndexOutOfBoundsException();
                }
            case 1:
                switch (row) {
                    case 0:
                        this.m10 = value;
                        return this;
                    case 1:
                        this.m11 = value;
                        return this;
                }
        }

        throw new IndexOutOfBoundsException();
    }

    public Matrix2f normal() {
        return this.normal(this);
    }

    public Matrix2f normal(Matrix2f dest) {
        float det = this.m00 * this.m11 - this.m10 * this.m01;
        float s = 1.0F / det;
        float nm00 = this.m11 * s;
        float nm01 = -this.m10 * s;
        float nm10 = -this.m01 * s;
        float nm11 = this.m00 * s;
        dest.m00 = nm00;
        dest.m01 = nm01;
        dest.m10 = nm10;
        dest.m11 = nm11;
        return dest;
    }

    public Vector2f getScale(Vector2f dest) {
        dest.x = Math.sqrt(this.m00 * this.m00 + this.m01 * this.m01);
        dest.y = Math.sqrt(this.m10 * this.m10 + this.m11 * this.m11);
        return dest;
    }

    public Vector2f positiveX(Vector2f dir) {
        if (this.m00 * this.m11 < this.m01 * this.m10) {
            dir.x = -this.m11;
            dir.y = this.m01;
        } else {
            dir.x = this.m11;
            dir.y = -this.m01;
        }

        return dir.normalize(dir);
    }

    public Vector2f normalizedPositiveX(Vector2f dir) {
        if (this.m00 * this.m11 < this.m01 * this.m10) {
            dir.x = -this.m11;
            dir.y = this.m01;
        } else {
            dir.x = this.m11;
            dir.y = -this.m01;
        }

        return dir;
    }

    public Vector2f positiveY(Vector2f dir) {
        if (this.m00 * this.m11 < this.m01 * this.m10) {
            dir.x = this.m10;
            dir.y = -this.m00;
        } else {
            dir.x = -this.m10;
            dir.y = this.m00;
        }

        return dir.normalize(dir);
    }

    public Vector2f normalizedPositiveY(Vector2f dir) {
        if (this.m00 * this.m11 < this.m01 * this.m10) {
            dir.x = this.m10;
            dir.y = -this.m00;
        } else {
            dir.x = -this.m10;
            dir.y = this.m00;
        }

        return dir;
    }

    public int hashCode() {
        int result = 1;
        result = 31 * result + Float.floatToIntBits(this.m00);
        result = 31 * result + Float.floatToIntBits(this.m01);
        result = 31 * result + Float.floatToIntBits(this.m10);
        result = 31 * result + Float.floatToIntBits(this.m11);
        return result;
    }

    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        } else if (obj == null) {
            return false;
        } else if (this.getClass() != obj.getClass()) {
            return false;
        } else {
            Matrix2f other = (Matrix2f) obj;
            if (Float.floatToIntBits(this.m00) != Float.floatToIntBits(other.m00)) {
                return false;
            } else if (Float.floatToIntBits(this.m01) != Float.floatToIntBits(other.m01)) {
                return false;
            } else if (Float.floatToIntBits(this.m10) != Float.floatToIntBits(other.m10)) {
                return false;
            } else {
                return Float.floatToIntBits(this.m11) == Float.floatToIntBits(other.m11);
            }
        }
    }

    public boolean equals(Matrix2f m, float delta) {
        if (this == m) {
            return true;
        } else if (m == null) {
            return false;
        } else if (!Runtime.equals(this.m00, m.m00, delta)) {
            return false;
        } else if (!Runtime.equals(this.m01, m.m01, delta)) {
            return false;
        } else if (!Runtime.equals(this.m10, m.m10, delta)) {
            return false;
        } else {
            return Runtime.equals(this.m11, m.m11, delta);
        }
    }

    public Matrix2f swap(Matrix2f other) {
        MemUtil.INSTANCE.swap(this, other);
        return this;
    }

    public Matrix2f add(Matrix2f other) {
        return this.add(other, this);
    }

    public Matrix2f add(Matrix2f other, Matrix2f dest) {
        dest.m00 = this.m00 + other.m00;
        dest.m01 = this.m01 + other.m01;
        dest.m10 = this.m10 + other.m10;
        dest.m11 = this.m11 + other.m11;
        return dest;
    }

    public Matrix2f sub(Matrix2f subtrahend) {
        return this.sub(subtrahend, this);
    }

    public Matrix2f sub(Matrix2f other, Matrix2f dest) {
        dest.m00 = this.m00 - other.m00;
        dest.m01 = this.m01 - other.m01;
        dest.m10 = this.m10 - other.m10;
        dest.m11 = this.m11 - other.m11;
        return dest;
    }

    public Matrix2f mulComponentWise(Matrix2f other) {
        return this.sub(other, this);
    }

    public Matrix2f mulComponentWise(Matrix2f other, Matrix2f dest) {
        dest.m00 = this.m00 * other.m00;
        dest.m01 = this.m01 * other.m01;
        dest.m10 = this.m10 * other.m10;
        dest.m11 = this.m11 * other.m11;
        return dest;
    }

    public Matrix2f lerp(Matrix2f other, float t) {
        return this.lerp(other, t, this);
    }

    public Matrix2f lerp(Matrix2f other, float t, Matrix2f dest) {
        dest.m00 = Math.fma(other.m00 - this.m00, t, this.m00);
        dest.m01 = Math.fma(other.m01 - this.m01, t, this.m01);
        dest.m10 = Math.fma(other.m10 - this.m10, t, this.m10);
        dest.m11 = Math.fma(other.m11 - this.m11, t, this.m11);
        return dest;
    }

    public boolean isFinite() {
        return Math.isFinite(this.m00) && Math.isFinite(this.m01) && Math.isFinite(this.m10) && Math.isFinite(this.m11);
    }

    public Object clone() throws CloneNotSupportedException {
        return super.clone();
    }
}

File: KOML/src/org/joml/Matrix2f.kt
package org.joml

import org.joml.JomlMath.addSigns
import org.joml.Runtime.f
import java.nio.FloatBuffer
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

@Suppress("unused")
open class Matrix2f : Matrix<Matrix2f, Vector2f, Vector2f> {

    var m00 = 0f
    var m01 = 0f
    var m10 = 0f
    var m11 = 0f

    constructor() {
        m00 = 1.0f
        m11 = 1.0f
    }

    constructor(mat: Matrix2f) {
        set(mat)
    }

    constructor(mat: Matrix3f) {
        set(mat)
    }

    constructor(m00: Float, m01: Float, m10: Float, m11: Float) {
        set(m00, m01, m10, m11)
    }

    constructor(col0: Vector2f, col1: Vector2f) {
        set(col0, col1)
    }

    override val numRows: Int get() = 2
    override val numCols: Int get() = 2

    fun _m00(m00: Float): Matrix2f {
        this.m00 = m00
        return this
    }

    fun _m01(m01: Float): Matrix2f {
        this.m01 = m01
        return this
    }

    fun _m10(m10: Float): Matrix2f {
        this.m10 = m10
        return this
    }

    fun _m11(m11: Float): Matrix2f {
        this.m11 = m11
        return this
    }

    fun set(m: Matrix2f): Matrix2f {
        return set(m.m00, m.m01, m.m10, m.m11)
    }

    fun set(m: Matrix3x2f): Matrix2f {
        return set(m.m00, m.m01, m.m10, m.m11)
    }

    fun set(m: Matrix3f): Matrix2f {
        return set(m.m00, m.m01, m.m10, m.m11)
    }

    fun putInto(arr: FloatBuffer): FloatBuffer {
        arr.put(m00).put(m01)
        arr.put(m10).put(m11)
        return arr
    }

    @JvmOverloads
    fun mul(right: Matrix2f, dst: Matrix2f = this): Matrix2f {
        return dst.set(
            m00 * right.m00 + m10 * right.m01,
            m01 * right.m00 + m11 * right.m01,
            m00 * right.m10 + m10 * right.m11,
            m01 * right.m10 + m11 * right.m11
        )
    }

    @JvmOverloads
    fun mulLocal(left: Matrix2f, dst: Matrix2f = this): Matrix2f {
        return dst.set(
            left.m00 * m00 + left.m10 * m01,
            left.m01 * m00 + left.m11 * m01,
            left.m00 * m10 + left.m10 * m11,
            left.m01 * m10 + left.m11 * m11
        )
    }

    fun set(m00: Float, m01: Float, m10: Float, m11: Float): Matrix2f {
        this.m00 = m00
        this.m01 = m01
        this.m10 = m10
        this.m11 = m11
        return this
    }

    fun set(m: FloatArray): Matrix2f {
        return set(m[0], m[1], m[2], m[3])
    }

    fun set(col0: Vector2f, col1: Vector2f): Matrix2f {
        return set(col0.x, col0.y, col1.x, col1.y)
    }

    fun determinant(): Float {
        return m00 * m11 - m10 * m01
    }

    @JvmOverloads
    fun invert(dst: Matrix2f = this): Matrix2f {
        val s = 1.0f / determinant()
        return dst.set(m11 * s, -m01 * s, -m10 * s, m00 * s)
    }

    @JvmOverloads
    fun transpose(dst: Matrix2f = this) = dst.set(m00, m10, m01, m11)

    override fun toString() =
        "[[${f(m00.toDouble())} ${f(m10.toDouble())}] [${f(m01.toDouble())} ${f(m11.toDouble())}]]".addSigns()

    fun get(dst: Matrix2f): Matrix2f {
        return dst.set(this)
    }

    fun get(dst: Matrix3x2f): Matrix3x2f {
        return dst.set(this)
    }

    fun get(dst: Matrix3f): Matrix3f {
        return dst.set(this)
    }

    val rotation: Float
        get() = atan2(m01, m11)

    fun get(arr: FloatArray, offset: Int = 0): FloatArray {
        arr[offset] = m00
        arr[offset + 1] = m01
        arr[offset + 2] = m10
        arr[offset + 3] = m11
        return arr
    }

    fun zero(): Matrix2f = set(0f, 0f, 0f, 0f)
    fun identity(): Matrix2f = set(1f, 0f, 0f, 1f)

    @JvmOverloads
    fun scale(xy: Vector2f, dst: Matrix2f = this): Matrix2f {
        return scale(xy.x, xy.y, dst)
    }

    @JvmOverloads
    fun scale(x: Float, y: Float, dst: Matrix2f = this): Matrix2f {
        return dst.set(m00 * x, m01 * x, m10 * y, m11 * y)
    }

    @JvmOverloads
    fun scale(xy: Float, dst: Matrix2f = this): Matrix2f {
        return this.scale(xy, xy, dst)
    }

    @JvmOverloads
    fun scaleLocal(x: Float, y: Float, dst: Matrix2f = this): Matrix2f {
        return dst.set(x * m00, y * m01, x * m10, y * m11)
    }

    fun scaling(x: Float, y: Float = x): Matrix2f {
        return set(x, 0f, 0f, y)
    }

    fun scaling(xy: Vector2f) = scaling(xy.x, xy.y)

    fun rotation(angle: Float): Matrix2f {
        val sin = sin(angle)
        val cos = cos(angle)
        return set(cos, sin, -sin, cos)
    }

    fun transform(v: Vector2f): Vector2f {
        return v.mul(this)
    }

    fun transform(v: Vector2f, dst: Vector2f): Vector2f {
        v.mul(this, dst)
        return dst
    }

    fun transform(x: Float, y: Float, dst: Vector2f): Vector2f {
        dst.set(m00 * x + m10 * y, m01 * x + m11 * y)
        return dst
    }

    fun transformTranspose(v: Vector2f): Vector2f {
        return v.mulTranspose(this)
    }

    fun transformTranspose(v: Vector2f, dst: Vector2f): Vector2f {
        v.mulTranspose(this, dst)
        return dst
    }

    fun transformTranspose(x: Float, y: Float, dst: Vector2f): Vector2f {
        dst.set(m00 * x + m01 * y, m10 * x + m11 * y)
        return dst
    }

    @JvmOverloads
    fun rotate(angle: Float, dst: Matrix2f = this): Matrix2f {
        val s = sin(angle)
        val c = cos(angle)
        return dst.set(
            m00 * c + m10 * s, m01 * c + m11 * s,
            m10 * c - m00 * s, m11 * c - m01 * s
        )
    }

    @JvmOverloads
    fun rotateLocal(angle: Float, dst: Matrix2f = this): Matrix2f {
        val s = sin(angle)
        val c = cos(angle)
        return dst.set(
            c * m00 - s * m01, s * m00 + c * m01,
            c * m10 - s * m11, s * m10 + c * m11
        )
    }

    override fun getRow(row: Int, dst: Vector2f): Vector2f {
        when (row) {
            0 -> dst.set(m00, m10)
            else -> dst.set(m01, m11)
        }
        return dst
    }

    override fun setRow(row: Int, src: Vector2f): Matrix2f {
        return setRow(row, src.x, src.y)
    }

    fun setRow(row: Int, x: Float, y: Float): Matrix2f {
        when (row) {
            0 -> {
                m00 = x
                m10 = y
            }
            else -> {
                m01 = x
                m11 = y
            }
        }
        return this
    }

    override fun getColumn(column: Int, dst: Vector2f): Vector2f {
        when (column) {
            0 -> dst.set(m00, m01)
            else -> dst.set(m10, m11)
        }
        return dst
    }

    override fun setColumn(column: Int, src: Vector2f): Matrix2f {
        return setColumn(column, src.x, src.y)
    }

    fun setColumn(column: Int, x: Float, y: Float): Matrix2f {
        when (column) {
            0 -> {
                m00 = x
                m01 = y
            }
            else -> {
                m10 = x
                m11 = y
            }
        }
        return this
    }

    override operator fun get(column: Int, row: Int): Double {
        return when (column * 2 + row) {
            0 -> m00
            1 -> m01
            2 -> m10
            else -> m11
        }.toDouble()
    }

    override operator fun set(column: Int, row: Int, value: Double): Matrix2f {
        return set(column, row, value.toFloat())
    }

    operator fun set(column: Int, row: Int, value: Float): Matrix2f {
        when (column * 2 + row) {
            0 -> m00 = value
            1 -> m01 = value
            2 -> m10 = value
            else -> m11 = value
        }
        return this
    }

    @JvmOverloads
    fun normal(dst: Matrix2f = this): Matrix2f {
        val det = m00 * m11 - m10 * m01
        val s = 1.0f / det
        return dst.set(
            m11 * s, -m10 * s,
            -m01 * s, m00 * s
        )
    }

    fun getScale(dst: Vector2f): Vector2f {
        return dst.set(
            sqrt(m00 * m00 + m01 * m01),
            sqrt(m10 * m10 + m11 * m11)
        )
    }

    fun positiveX(dir: Vector2f): Vector2f {
        if (m00 * m11 < m01 * m10) {
            dir.x = -m11
            dir.y = m01
        } else {
            dir.x = m11
            dir.y = -m01
        }
        return dir.normalize(dir)
    }

    fun normalizedPositiveX(dir: Vector2f): Vector2f {
        if (m00 * m11 < m01 * m10) {
            dir.x = -m11
            dir.y = m01
        } else {
            dir.x = m11
            dir.y = -m01
        }
        return dir
    }

    fun positiveY(dir: Vector2f): Vector2f {
        if (m00 * m11 < m01 * m10) {
            dir.x = m10
            dir.y = -m00
        } else {
            dir.x = -m10
            dir.y = m00
        }
        return dir.normalize(dir)
    }

    fun normalizedPositiveY(dir: Vector2f): Vector2f {
        if (m00 * m11 < m01 * m10) {
            dir.x = m10
            dir.y = -m00
        } else {
            dir.x = -m10
            dir.y = m00
        }
        return dir
    }

    override fun hashCode(): Int {
        var result = 1
        result = 31 * result + m00.toRawBits()
        result = 31 * result + m01.toRawBits()
        result = 31 * result + m10.toRawBits()
        result = 31 * result + m11.toRawBits()
        return result
    }

    override fun equals(other: Any?): Boolean {
        return other is Matrix2f && m00 == other.m00 && m01 == other.m01 && m10 == other.m10 && m11 == other.m11
    }

    override fun equals(other: Matrix2f?, threshold: Double): Boolean {
        return equals(other, threshold.toFloat())
    }

    fun equals(m: Matrix2f?, delta: Float): Boolean {
        return m != null &&
                Runtime.equals(m00, m.m00, delta) && Runtime.equals(m01, m.m01, delta) &&
                Runtime.equals(m10, m.m10, delta) && Runtime.equals(m11, m.m11, delta)
    }

    @JvmOverloads
    fun add(other: Matrix2f, dst: Matrix2f = this): Matrix2f {
        return dst.set(
            m00 + other.m00, m01 + other.m01,
            m10 + other.m10, m11 + other.m11
        )
    }

    @JvmOverloads
    fun sub(other: Matrix2f, dst: Matrix2f = this): Matrix2f {
        return dst.set(
            m00 - other.m00, m01 - other.m01,
            m10 - other.m10, m11 - other.m11
        )
    }

    @JvmOverloads
    fun mulComponentWise(other: Matrix2f, dst: Matrix2f = this): Matrix2f {
        return dst.set(
            m00 * other.m00, m01 * other.m01,
            m10 * other.m10, m11 * other.m11
        )
    }

    @JvmOverloads
    fun mix(other: Matrix2f, t: Float, dst: Matrix2f = this): Matrix2f {
        return dst.set(
            (other.m00 - m00) * t + m00, (other.m01 - m01) * t + m01,
            (other.m10 - m10) * t + m10, (other.m11 - m11) * t + m11
        )
    }

    @JvmOverloads
    fun lerp(other: Matrix2f, t: Float, dst: Matrix2f = this): Matrix2f {
        return mix(other, t, dst)
    }

    val isFinite: Boolean
        get() = JomlMath.isFinite(m00) && JomlMath.isFinite(m01) &&
                JomlMath.isFinite(m10) && JomlMath.isFinite(m11)
}